cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -602,52 +828,155 @@ class AsofJoinNode : public ExecNode {
 
  public:
   AsofJoinNode(ExecPlan* plan, NodeVector inputs, std::vector<std::string> input_labels,
-               const AsofJoinNodeOptions& join_options,
+               const vec_col_index_t& indices_of_on_key,
+               const std::vector<vec_col_index_t>& indices_of_by_key, OnType tolerance,
                std::shared_ptr<Schema> output_schema);
 
+  Status InternalInit(bool must_hash, bool nullable_by_key,
```

Comment:
`InternalInit` is a bit of an odd name.  It sort of looks like the `Make` method was somewhat arbitrarily split in two.  Did you intend to call this somewhere else too?

Reply:
I needed an initialization method that returns `Status` in order to deal with status-errors correctly, which cannot be done in a constructor. I tried the name `Init` but that resulted in a compile error due to a conflict with an `Init` method in a super-class.

Reply:
Actually, the method can be made void, which I will do shortly. There is inter-dependency between `AsofJoinNode` and `KeyHasher` which prevents full initialization using constructors, so `InternalInit` completes the initialization.

Reply:
I figured out a way to clean this up and will commit a fix shortly.